ui/pages/settings/index.js

```diff
@@ -675,19 +675,138 @@ function Admins() {
   )
 }
 
-function Providers() {
+function Authentication() {
   const { data: { items: providers } = {} } = useSWR(
     `/api/providers?&limit=1000`
   )
+  const { data: org } = useSWR('/api/organizations/self')
+  const [allowedDomains, setAllowedDomains] = useState(
+    org?.allowedDomains || []
+  )
+  const [newDomain, setNewDomain] = useState('')
+  const [error, setError] = useState('')
+
+  async function removeDomain(e) {
+    e.preventDefault()
+    let toRemove = e.target.value
+    const newAllowedDomains = allowedDomains.filter(d => d !== toRemove)
+    updateAllowedDomains(newAllowedDomains)
+  }
+
+  async function addDomain(e) {
+    e.preventDefault()
+    // check that we only have the domain (no protocol), but be lenient on validation
+    let cleanedInput = newDomain
+    if (newDomain.startsWith('http://')) {
+      cleanedInput = cleanedInput.replace('http://', '')
+    }
+    if (newDomain.startsWith('https://')) {
+      cleanedInput = cleanedInput.replace('https://', '')
+    }
+    if (newDomain.startsWith('www.')) {
+      cleanedInput = cleanedInput.replace('www.', '')
+    }
+    if (newDomain.startsWith('@')) {
+      cleanedInput = cleanedInput.replace('@', '')
+    }
+    if (!allowedDomains.includes(cleanedInput)) {
+      const newAllowedDomains = [...allowedDomains, cleanedInput]
+      updateAllowedDomains(newAllowedDomains)
+    }
+    setNewDomain('')
+  }
+
+  function onKeyDown(e) {
+    const { key } = e
+
+    if (key === 'Backspace' && newDomain === '' && allowedDomains.length > 0) {
+      e.preventDefault()
+      const newAllowedDomains = [...allowedDomains]
+      newAllowedDomains.pop()
+      updateAllowedDomains(newAllowedDomains)
+    }
+  }
+
+  async function updateAllowedDomains(allowedDomains) {
+    setError('')
+    try {
+      const res = await fetch('/api/organizations/' + org.id, {
+        method: 'PUT',
+        body: JSON.stringify({ allowedDomains }),
+      })
+      await jsonBody(res)
+      setAllowedDomains(allowedDomains)
+      mutate('/api/organizations/self')
+    } catch (e) {
+      setError(e.message)
+    }
+  }
 
   return (
     <>
-      <header className='my-2 flex items-center justify-end'>
+      {providers?.some(p => p.id === googleSocialLoginID) && (
+        <>
+          <header className='my-6 flex flex-col justify-between md:flex-row md:space-y-0 md:space-x-4'>
+            <div>
+              <h2 className='mb-0.5 flex items-center font-display text-lg font-medium'>
+                Allowed Domains
+              </h2>
+              <h3 className='text-sm text-gray-500'>
+                Google accounts with these domains are able to log in to your
+                organization. They will not have any infrastructure access by
+                default.
```

Comment:
Optional: the second sentence seems like the default expectation for the user and can be left out

```suggestion
              <h3 className='text-sm text-gray-500'>
                Users with these domains are able to log in to your
                organization.
```